MaxText/layers/linears.py

```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
     return wrapper(inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel)
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
+    # shape of top_k_weights & top_k_indices: (batch, sequence, exp_per_token)
     top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
-    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
-
-    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
-    softmax_probs = softmax_probs.reshape(gate_logits.shape[:-1] + (self.num_experts_per_tok,))
+    softmax_probs = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
+
+    # token dropping
+    if self.config.capacity_factor > 0:
```

Comment:
Can we wrap the body of this in its own method - for now the method just needs to return the softmax_probs, perhaps the mask as well once we integrate this will Expert parallelism

Reply:
Sg!

Reply:
Moved to generate_masks() helper method.